students/mitch334/lesson09/mailroom.py

```diff
@@ -0,0 +1,244 @@
+# Update your mailroom program to:
+# * Use dicts where appropriate.
+# * See if you can use a dict to switch between the users selections. see 'Using a Dictionary to switch' for what this means.
+# * Convert your main donor data structure to be a dict.
+# * Try to use a dict and the .format() method to produce the letter as one big template, rather than building up a big string that produces the letter in parts.
+
+# The Program: Part 3
+# Exceptions
+# Now that you’ve learned about exception handling, you can update your code to handle errors better, such as when a user inputs bad data.
+# Comprehensions
+# Can you use comprehensions to clean up your code a bit?
+
+# Add a full suite of unit tests.
+# The Program: Part 4
+# “Full suite” means all the code is tested. In practice, it’s very hard to test the user interaction, but you can test everything else. Therefore you should make sure that there is as little logic and untested code in the user interaction portion of the program as possible.
+# This is a big step; you may find that your code is hard to test. If that’s the case, it’s a good sign that you should refactor your code.
+# I like to say: “If it’s hard to test, it’s not well structured.”
+# Put in the tests before you make the other changes below. That’s much of the point of tests. You can know that you haven’t broken anything when you refactor!
+
+# The Program: Part 5
+# Goal: Refactor the mailroom program using classes to help organize the code.
+
+
+import os
+import tempfile
+import datetime
+
+class Donor:
+    """Class responsible for donor data encapsulation"""
+
+    def __init__(self, name, donations=None):
+        self.name = name
+
+        if donations == None:
+            self._donations = []
+        else:
+            self._donations = list(donations)
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        self._name = value
+
+    @property
+    def donations(self):
+        return self._donations
+
+    def add_donation(self, amount):
+        self.donations.append(amount)
+
+    def donations_sum(self):
+        return sum(self.donations)
+
+    def donations_count(self):
+        return len(self.donations)
+
+    def donations_avg(self):
+        return self.donations_sum() / self.donations_count()
+
+    def create_email(self, amount):
+        """Create emails after donation"""
+
+        return (f'Dear {self.name},\n\nThank you for the generous donation of ${amount:.2f}.\n\n'
+            'Sincerely,\nMatthew Mitchell')
+
+
+class DonorCollection:
+    """Class responsible for donor collection data encapsulation"""
+
+    def __init__(self, donors=None):
+        if donors == None:
+            donors = []
+        else:
+            self._donors = donors
+
+    @property
+    def donors(self):
+        return self._donors
+
+    def donors_list(self, list_output):
+        """Create donor list or database details"""
+
+        list = '\n'
```

Comment:
you can't use `list` as a variable name - you are breaking built-in `list()` function